### PR TITLE
Adds stable layout types to pass to the runtime

### LINKED ISSUE #29852: Audit types used by VM due to rust 1.68 memory layout changes
#### Problem

Rust 1.68 has [changed memory layout of certain types](https://github.com/rust-lang/rust/pull/102750), which [impacts `Vec`](https://godbolt.org/z/ojz6fxacM) and other types within the codebase. This is an issue for any types that are used within the VM where its memory layout is assumed.

This was discovered when [upgrading Rust to 1.66.0](https://github.com/solana-labs/solana/pull/29304), which also upgraded `nightly` to 1.68. The [`check_type_assumptions` test failed](https://buildkite.com/solana-labs/solana/builds/87597#0185824a-16d0-4600-b184-dd25e769ad9c) due to the layout of `Instruction` and `AccountInfo` changing.
https://github.com/solana-labs/solana/blob/aef1a4301cd0f57bc274f5706b14f1e365c6426c/sdk/program/src/program.rs#L394-L545

For types that are used that assume a certain layout, we may need to create our own stable-layout versions of those types. Here's an example for `Vec`:
https://godbolt.org/z/v4qa9vGYz

I noticed at least a few types in the VM do this already with types prefixed by `Sol`, i.e. `SolInstruction`, `SolAccountMeta`, and `SolAccountInfo`:
https://github.com/solana-labs/solana/blob/aef1a4301cd0f57bc274f5706b14f1e365c6426c/programs/bpf_loader/src/syscalls/cpi.rs#L390-L425

#### Proposed Solution

Audit types used by the VM that may assume layout (`Instruction`, `AccountInfo`, `AccountMeta`, etc) and see if any require a stable layout. Nail down those layouts and update the `check_type_assumptions` test.

### PR DESCRIPTION
**Pull request by brooksprumo**
#### Problem

Please refer to https://github.com/solana-labs/solana/issues/29852.

#### Summary of Changes

Add shadow structs with stable memory layouts to pass to the runtime.

Fixes #29852

#### Testing

In addition to CI, I built `solana-validator` with Rust nightly (`cargo 1.69.0-nightly (e84a7928d 2023-01-31)`) and ran it against mainnet-beta. It has caught up and has been running & making roots for 21+ hours.

Additionally, I've run `cargo +nightly test` against the following dirs:
* sdk/
* sdk/cargo-build-sbf/
* sdk/cargo-build-bpf/
* sdk/cargo-test-sbf/
* sdk/cargo-test-bpf/
* sdk/program/
* programs/bpf_loader/
* programs/bpf-loader_tests/
* program-runtime/
* program-test/

### CODE REVIEW

File: program-test/src/lib.rs [hunk 1]
```diff
@@ -223,7 +224,7 @@ impl solana_sdk::program_stubs::SyscallStubs for SyscallStubs {
 
     fn sol_invoke_signed(
         &self,
-        instruction: &Instruction,
+        instruction: &StableInstruction,
```
Comment (alessandrod) [inline program-test/src/lib.rs:227]: this is breaking downstream-projects. This stubs stuff is public API so I think you'll have
to create the stable value internally

File: sdk/program/src/stable_layout.rs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+//! bprumo TODO: doc, or hide doc?
+
```
Comment (alessandrod) [inline sdk/program/src/stable_layout.rs:2]: /// DO NOT USE THIS MODULE. FOR INTERNAL USE ONLY. IF YOU USE THIS MODULE ALL
/// YOUR APES GONE.

File: sdk/program/src/stable_layout/stable_instruction.rs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        let instruction = Instruction {
+            program_id,
+            accounts: accounts.clone(),
+            data: data.clone(),
+        };
+        let instruction = StableInstruction::from(instruction);
+
+        let instruction_addr = &instruction as *const _ as u64;
+
+        let accounts_ptr = instruction_addr as *const &[AccountMeta; 2];
```
Comment (alessandrod) [inline sdk/program/src/stable_layout/stable_instruction.rs:68]: Shouldn't this cast to *const StableVec?

File: sdk/program/src/stable_layout/stable_instruction.rs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        assert_eq!(unsafe { *accounts_ptr }, accounts.as_slice());
+
+        let data_ptr = (instruction_addr + 24) as *const &[u8; 5];
```
Comment (alessandrod) [inline sdk/program/src/stable_layout/stable_instruction.rs:71]: Same here, I think you're getting lucky that len == cap (vec![] sets capacity to
the n of elements passed to the macro), so it accidentally matches the slice

**Review by alessandrod — COMMENTED** [review]
Solid PR so far! Just a couple of minor comments but otherwise looks great

File: sdk/program/src/program.rs [hunk 5]
```diff
@@ -292,9 +293,11 @@ pub fn invoke_signed_unchecked(
 ) -> ProgramResult {
     #[cfg(target_os = "solana")]
     {
+        // bprumo TODO: it would be nice to not clone `instruction`... Can this fn take a StableInstruction instead?
+        let instruction = StableInstruction::from(instruction.clone());
```
[the remark this answers is not in the document]
Comment (alessandrod) [reply sdk/program/src/program.rs:297]: That clone there doesn't really matter - that code path is only executed during tests when not actually exercising the program runtime. So perf doesn't matter, and it's a fringe case anyway.

File: sdk/program/src/stable_layout/stable_vec.rs [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+/// `Vec`, with a stable memory layout
+///
+/// This container is used within the runtime to ensure memory mapping and memory accesses are
+/// valid.  We rely on known addresses and offsets within the runtime, and since `Vec`'s layout is
+/// allowed to change, we must provide a way to lock down the memory layout.  `StableVec`
+/// reimplements the bare minimum of `Vec`'s API sufficient only for the runtime's needs.
+///
+/// To ensure memory allocation and deallocation is handled correctly, it is only possible to
+/// create a new `StableVec` from an existing `Vec`.  This way we ensure all Rust invariants are
+/// upheld.
```
[the remark this answers is not in the document]
Comment (alessandrod) [reply sdk/program/src/stable_layout/stable_vec.rs:14]: Yep I saw this it's perfect as it is. We only unpack vecs we never modify them (so they shouldn't be vecs in the first place, but... compatibility).

File: sdk/program/src/program.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (alessandrod) [reply sdk/program/src/program.rs]: Err brain fart, somehow I thought the conversion was only for the stub call but obvs it's needed for the actual syscall too. It's still an infrequent operation tho so it's fine

File: programs/address-lookup-table/src/processor.rs [hunk 7]
```diff
@@ -144,18 +145,25 @@ impl Processor {
 
         if required_lamports > 0 {
             invoke_context.native_invoke(
-                system_instruction::transfer(&payer_key, &table_key, required_lamports),
+                StableInstruction::from(system_instruction::transfer(
+                    &payer_key,
+                    &table_key,
+                    required_lamports,
+                )),
```
[the remark this answers is not in the document]
Comment (alessandrod) [reply programs/address-lookup-table/src/processor.rs:152]: Yup the assumption is correct. Although in this case I think using `.into()` would be better

File: sdk/program/src/stable_layout/stable_vec.rs [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+        // correctly, convert ourselves back to a Vec and let Vec's drop handling take over.
+        //
+        // SAFETY: We have a valid StableVec, which we can only get from a Vec.  Therefore it is
+        // safe to convert back to Vec.
+        let _vec = unsafe { Vec::from_raw_parts(self.ptr.as_ptr(), self.len, self.cap) };
+
+        // For extra piece of mind, zero-out our guts
+        self.ptr = NonNull::dangling();
+        self.cap = 0;
+        self.len = 0;
```
[the remark this answers is not in the document]
Comment (alessandrod) [reply sdk/program/src/stable_layout/stable_vec.rs:147]: Yeah this is unnecessary

File: sdk/src/lib.rs [hunk 9]
```diff
@@ -50,8 +50,8 @@ pub use solana_program::{
     instruction, keccak, lamports, loader_instruction, loader_upgradeable_instruction, message,
     msg, native_token, nonce, program, program_error, program_memory, program_option, program_pack,
     rent, sanitize, sdk_ids, secp256k1_program, secp256k1_recover, serde_varint, serialize_utils,
-    short_vec, slot_hashes, slot_history, stake, stake_history, syscalls, system_instruction,
-    system_program, sysvar, unchecked_div_by_const, vote, wasm_bindgen,
+    short_vec, slot_hashes, slot_history, stable_layout, stake, stake_history, syscalls,
```
[the remark this answers is not in the document]
Comment (alessandrod) [reply sdk/src/lib.rs:53]: Yeah we don't have many options do we? Either this, or we do hacks using include_bytes or #[path=...] etc, which seem worse. 

We did document it's for internal use only 🤷‍♂️

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 10]
```diff
@@ -296,10 +298,12 @@ impl SyscallInvokeSigned for SyscallInvokeSignedRust {
             invoke_context.get_check_size(),
         )?
         .to_vec();
-        Ok(Instruction {
-            program_id: ix.program_id,
+        let data = StableVec::from(data);
```
Comment (alessandrod) [inline programs/bpf_loader/src/syscalls/cpi.rs:301]: this could also be data: data.into() below

File: programs/bpf_loader/src/syscalls/cpi.rs [hunk 11]
```diff
@@ -514,6 +518,8 @@ impl SyscallInvokeSigned for SyscallInvokeSignedC {
             invoke_context.get_check_size(),
         )?
         .to_vec();
+        let data = StableVec::from(data);
```
Comment (alessandrod) [inline programs/bpf_loader/src/syscalls/cpi.rs:521]: and here

File: sdk/program/src/stable_layout/stable_vec.rs [hunk 12]
```diff
@@ -0,0 +1,186 @@
...
+    }
+}
+
+impl<T> AsMut<[T]> for StableVec<T> {
+    fn as_mut(&mut self) -> &mut [T] {
+        self
+    }
+}
+
+impl<T> std::ops::Deref for StableVec<T> {
```
Comment (alessandrod) [inline sdk/program/src/stable_layout/stable_vec.rs:61]: I don't think we're using these extra trait impls, and if so I'd say let's
remove them

**Comment by alessandrod** [comment]
> $ rg translate_type | sed -n 's/.*translate_type.*::<\(.*\)>.*/\1/p' | sort | uniq
[snip]
> **T**
[snip]

😱

Looks like that scary T is `get_sysvar`, so all sysvars need to be checked (I checked a couple and they're repr(C) so hopefully they all are!

File: sdk/program/src/stable_layout/stable_vec.rs (hunk 12 above)
Comment (alessandrod) [reply sdk/program/src/stable_layout/stable_vec.rs:61]: Ah uh no this seems worse, I say leave it as is

**Review by alessandrod — APPROVED** [review]
This looks good to me now, thanks for much for doing this work!

Up to you if you want to try to hack downstream-projects in CI to run with nightly, but if a validator running this makes roots, I think we're probably good to land!

**Comment by alessandrod** [comment]
> I _think_ we're OK, since bpf_loader only cares about Clock, EpochSchedule, Fees, and Rent. Those are the only sysvars that I saw that had Syscalls associated with them. which I think means those are the only ones that use `translate_type()`?

Yep the other sysvars use bincode (eew), so we're good

**Comment by joncinque** [comment]
The lack of this in 1.14 is causing issues for people who build their on-chain program with the compiler from 1.16 and the 1.14 crates. You can see my writeup at https://github.com/solana-labs/solana/issues/31960#issuecomment-1668682153

Although we don't typically recommend that people do that, the other difficulties with upgrading to the 1.16 crates has meant that people first upgrade their toolchain, only to find that their current program doesn't work.

Would it be possible to backport this PR to 1.14? I know it's very late, but the changes don't seem too invasive, and it'll unblock a lot of developers.

**Pull request merged**